source/agora/test/PreImageCatchup.d

```diff
@@ -0,0 +1,64 @@
+/*******************************************************************************
+
+    Check that a validator can retrieve missing pre-images when it could not
+    receive pre-images through `postPreimage`. We altered the algorithm of
+    distributing pre-images so that a validator does not continuously send
+    pre-images any more without checking that the previous pre-image is
+    posted normally. So the network should survive the situation where
+    some pre-images are missing.
+
+    Copyright:
+        Copyright (c) 2019-2022 BOSAGORA Foundation
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.test.PreImageCatchup;
+
+version (unittest):
+
+import agora.consensus.data.PreImageInfo;
+import agora.test.Base;
+
+/// A validator should retrive the pre-images even when the `postPreimage`
+/// does not work or the validator misses some pre-images.
+unittest
+{
+    static class CustomValidator : TestValidatorNode
+    {
+        mixin ForwardCtor!();
+
+        /// It makes the situation where the node doesn't receive pre-images
+        /// and miss some preimages from some point.
+        public override bool postPreimage (in PreImageInfo preimage) @safe
+        {
+            if (this.ledger.getBlockHeight() <= 1)
+                return super.postPreimage(preimage);
+            return true;
+        }
+    }
```

Comment:
What is this testing ? The golden rule of testing is that your test should fail if the fix is not present.
From the look of it, this test would pass even without the fix.

Reply:
Modified the test based on the comment.